src/test/java/uk/nhs/hee/tis/trainee/sync/CachingIntTest.java

```diff
@@ -0,0 +1,154 @@
+import java.util.Optional;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.cache.Cache;
+import org.springframework.cache.CacheManager;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Primary;
+import uk.nhs.hee.tis.trainee.sync.config.MongoConfiguration;
+import uk.nhs.hee.tis.trainee.sync.model.Operation;
+import uk.nhs.hee.tis.trainee.sync.model.Post;
+import uk.nhs.hee.tis.trainee.sync.repository.PostRepository;
+import uk.nhs.hee.tis.trainee.sync.service.PostSyncService;
+
+
+@SpringBootTest
+class CachingIntTest {
+
+  private static final String POST_FORDY = "fordy";
+
+  // We require access to the mock before the proxy wraps it.
+  private static PostRepository mPostRepository;
+
+  @Autowired
+  PostSyncService postSyncService;
+
+  @Autowired
+  CacheManager cacheManager;
+
+  private Cache postCache;
+
+  private Post post;
+
+  @BeforeEach
+  void setup() {
+    post = new Post();
+    post.setTisId(POST_FORDY);
+    post.setOperation(Operation.DELETE);
+    post.setTable(Post.ENTITY_NAME);
+
+    postCache = cacheManager.getCache(Post.ENTITY_NAME);
+  }
+
+  @Test
+  void shouldHitCacheOnSecondInvocation() {
+    when(mPostRepository.findById(POST_FORDY))
+        .thenReturn(Optional.of(post), Optional.of(new Post()));
+    assertThat(postCache.get(POST_FORDY)).isNull();
+
+    Optional<Post> actual1 = postSyncService.findById(POST_FORDY);
+    assertThat(postCache.get(POST_FORDY)).isNotNull();
+    Optional<Post> actual2 = postSyncService.findById(POST_FORDY);
+
+    verify(mPostRepository).findById(POST_FORDY);
+    assertThat(actual1).isPresent().get().isEqualTo(post).isEqualTo(actual2.orElseThrow());
+  }
+
+  @Test
+  void shouldEvictWhenDeletedAndMissCacheOnNextInvocation() {
+    Post otherPost = new Post();
+    final String otherKey = "Foo";
+    otherPost.setTisId(otherKey);
+    otherPost.setOperation(Operation.LOAD);
+    when(mPostRepository.findById(otherKey)).thenReturn(Optional.of(otherPost));
+    postSyncService.findById(otherKey);
+    assertThat(postCache.get(otherKey)).isNotNull();
+    when(mPostRepository.findById(POST_FORDY)).thenReturn(Optional.of(post));
+    postSyncService.findById(POST_FORDY);
+    assertThat(postCache.get(POST_FORDY)).isNotNull();
+
+    postSyncService.syncRecord(post);
+    assertThat(postCache.get(POST_FORDY)).isNull();
+    assertThat(postCache.get(otherKey)).isNotNull();
+    verify(mPostRepository).deleteById(POST_FORDY);
+
+    postSyncService.findById(POST_FORDY);
+    assertThat(postCache.get(POST_FORDY)).isNotNull();
+
+    verify(mPostRepository, times(2)).findById(POST_FORDY);
```

Comment:
From what I saw online it looks like the counter of interactions of a mock isn't automatically wiped after a verification or perhaps even between tests, so it might fail and count 3 interactions rather than 2 because its picking up other previous interactions (?). The test passes individually but not when all tests are run collectively.

It seems to work if we use:
```
@AfterEach
void teardown() {
  Mockito.reset(mPostRepository);
}
```
so that the mock starts anew with every next test (I believe `reset()` will reset its counter and its stubbed methods too), kinda like reinstantiating it but without reinstantiating it.

Reply:
As above, `@DirtiesContext` makes sure the context isn't re-used to give better test isolation.